src/calcTxnHash_nbgl.c

```diff
@@ -0,0 +1,243 @@
+    content->tagValueList.callback = NULL;
+
+    content->tagValueList.startIndex = 0;
+    content->tagValueList.wrapping = false;
+    content->tagValueList.smallCaseForValue = false;
+    content->tagValueList.nbMaxLinesForValue = 0;
+
+    return true;
+}
+
+static void zero_ctx(void) {
+    explicit_bzero(ctx, sizeof(calcTxnHashContext_t));
+}
+
+// handleCalcTxnHash reads a signature index and a transaction, calculates the
+// SigHash of the transaction, and optionally signs the hash using a specified
+// key. The transaction is processed in a streaming fashion and displayed
+// piece-wise to the user.
+void handleCalcTxnHash(uint8_t p1, uint8_t p2, uint8_t *dataBuffer, uint16_t dataLength, volatile unsigned int *flags, volatile unsigned int *tx __attribute__((unused))) {
+    if ((p1 != P1_FIRST && p1 != P1_MORE) || (p2 != P2_DISPLAY_HASH && p2 != P2_SIGN_HASH)) {
+        THROW(SW_INVALID_PARAM);
+    }
+
+    if (p1 == P1_FIRST) {
+        // If this is the first packet of a transaction, the transaction
+        // context must not already be initialized. (Otherwise, an attacker
+        // could fool the user by concatenating two transactions.)
+        //
+        // NOTE: ctx->initialized is set to false when the Sia app loads.
+        if (ctx->initialized) {
+            zero_ctx();
+            THROW(SW_IMPROPER_INIT);
+        }
+        explicit_bzero(ctx, sizeof(calcTxnHashContext_t));
+        ctx->initialized = true;
+
+        // If this is the first packet, it will include the key index, sig
+        // index, and change index in addition to the transaction data. Use
+        // these to initialize the ctx and the transaction decoder.
+        ctx->keyIndex = U4LE(dataBuffer, 0);  // NOTE: ignored if !ctx->sign
+        dataBuffer += 4;
+        dataLength -= 4;
+        uint16_t sigIndex = U2LE(dataBuffer, 0);
+        dataBuffer += 2;
+        dataLength -= 2;
+        uint32_t changeIndex = U4LE(dataBuffer, 0);
+        dataBuffer += 4;
+        dataLength -= 4;
+        txn_init(&ctx->txn, sigIndex, changeIndex);
+
+        // Set ctx->sign according to P2.
+        ctx->sign = (p2 & P2_SIGN_HASH);
+
+        ctx->elemPart = 0;
+    } else {
+        // If this is not P1_FIRST, the transaction must have been
+        // initialized previously.
+        if (!ctx->initialized) {
+            zero_ctx();
+            THROW(SW_IMPROPER_INIT);
+        }
+    }
+
+    // Add the new data to transaction decoder.
+    txn_update(&ctx->txn, dataBuffer, dataLength);
+
+    switch (txn_parse(&ctx->txn)) {
+        case TXN_STATE_ERR:
+            // don't leave state lingering
+            zero_ctx();
+            THROW(SW_INVALID_PARAM);
+            break;
+        case TXN_STATE_PARTIAL:
+            THROW(SW_OK);
+            break;
+        case TXN_STATE_FINISHED:
+            *flags |= IO_ASYNCH_REPLY;
+            nbgl_useCaseRegularReview(0, 0, "Cancel", NULL, nav_callback, confirm_callback);
```

Comment:
Any reason not to use the classical mecanism with `nbgl_useCaseReviewStart` + `nbgl_useCaseStaticReview` here?
Note that `nbgl_useCaseStaticReview`  support a mode where pairs are passed via a callback, so that you don't have to load them all on the RAM at the same time if needed.

Reply:
Ok I got it, you should still use the `nbgl_useCaseReviewStart` first, but I get why you are using `nbgl_useCaseRegularReview` after that. Thought the number of page should be set as second parameter?